x/ccv/provider/keeper/consumer_lifecycle.go

```diff
@@ -0,0 +1,670 @@
+				"error", err)
+			continue
+		}
+		if removalTime.After(ctx.BlockTime()) {
+			return result
+		}
+
+		consumers, err := k.GetConsumersToBeRemoved(ctx, removalTime)
+		if err != nil {
+			k.Logger(ctx).Error("failed to retrieve consumers to remove",
+				"removal time", removalTime,
+				"error", err)
+			continue
+		}
+		if len(result)+len(consumers.Ids) >= int(limit) {
+			remainingConsumerIds := len(result) + len(consumers.Ids) - int(limit)
+			if len(consumers.Ids[:len(consumers.Ids)-remainingConsumerIds]) == 0 {
+				return result
+			}
+			return append(result, consumers.Ids[:len(consumers.Ids)-remainingConsumerIds]...)
+		} else {
+			result = append(result, consumers.Ids...)
+		}
+	}
+
+	return result
+}
+
+// DeleteConsumerChain cleans up the state of the given consumer chain
+func (k Keeper) DeleteConsumerChain(ctx sdk.Context, consumerId string) (err error) {
+	phase := k.GetConsumerPhase(ctx, consumerId)
+	if phase != types.CONSUMER_PHASE_STOPPED {
+		return fmt.Errorf("cannot delete non-stopped chain: %s", consumerId)
+	}
+
+	// clean up states
+	k.DeleteConsumerClientId(ctx, consumerId)
+	k.DeleteConsumerGenesis(ctx, consumerId)
+	// Note: this call panics if the key assignment state is invalid
+	k.DeleteKeyAssignments(ctx, consumerId)
+	k.DeleteMinimumPowerInTopN(ctx, consumerId)
+	k.DeleteEquivocationEvidenceMinHeight(ctx, consumerId)
+
+	// close channel and delete the mappings between chain ID and channel ID
+	if channelID, found := k.GetConsumerIdToChannelId(ctx, consumerId); found {
+		// Close the channel for the given channel ID on the condition
+		// that the channel exists and isn't already in the CLOSED state
+		channel, found := k.channelKeeper.GetChannel(ctx, ccv.ProviderPortID, channelID)
+		if found && channel.State != channeltypes.CLOSED {
+			err := k.chanCloseInit(ctx, channelID)
+			if err != nil {
+				k.Logger(ctx).Error("channel to consumer chain could not be closed",
+					"consumerId", consumerId,
+					"channelID", channelID,
+					"error", err.Error(),
+				)
+			}
+		}
+		k.DeleteConsumerIdToChannelId(ctx, consumerId)
+		k.DeleteChannelIdToConsumerId(ctx, channelID)
+	}
+
+	// delete consumer commission rate
+	provAddrs := k.GetAllCommissionRateValidators(ctx, consumerId)
+	for _, addr := range provAddrs {
+		k.DeleteConsumerCommissionRate(ctx, consumerId, addr)
+	}
+
+	k.DeleteInitChainHeight(ctx, consumerId)
+	k.DeleteSlashAcks(ctx, consumerId)
+	k.DeletePendingVSCPackets(ctx, consumerId)
+
+	k.DeleteAllowlist(ctx, consumerId)
+	k.DeleteDenylist(ctx, consumerId)
+	k.DeleteAllOptedIn(ctx, consumerId)
+	k.DeleteConsumerValSet(ctx, consumerId)
+
+	k.DeleteConsumerRewardsAllocation(ctx, consumerId)
+	k.DeleteConsumerRemovalTime(ctx, consumerId)
+
+	// TODO (PERMISSIONLESS) add newly-added state to be deleted
+	// Note that we do not delete ConsumerIdToChainIdKey and ConsumerIdToPhase, as well
+	// as consumer metadata, initialization and power-shaping parameters.
```

Comment:
Please add some minimal explanation of why we don't delete this here

Reply:
see https://github.com/cosmos/interchain-security/pull/2237/commits/1ab477bc7061e89f51cac19fed9d287f12c630bf